src/core/algorithms/dc/verifier/dc_verifier.cpp

```diff
@@ -0,0 +1,521 @@
+        if (!res) return false;
+    }
+
+    return true;
+}
+
+void DCVerifier::ProcessMixed(std::vector<dc::Predicate> const& preds, Tree& insert_tree,
+                              Tree const& search_tree, dc::DC const& dc, size_t i,
+                              std::vector<Column::IndexType> const& cols_indices, bool& res) {
+    std::vector<std::byte const*> tuple = GetTuple(i);
+    auto const [t_box, t_inv_box] = SearchRanges(dc, tuple);
+    if (Eval(tuple, preds)) {
+        std::vector<Point> search_res = search_tree.QuerySearch(t_box);
+        std::vector<Point> inv_search_res = search_tree.QuerySearch(t_inv_box);
+
+        if (!search_res.empty() or !inv_search_res.empty()) {
+            res = false;
+            return;
+        };
+
+        insert_tree.Insert(MakePoint(tuple, cols_indices, i + index_offset_));
+    }
+
+    res = true;
+    return;
+}
+
+bool DCVerifier::VerifyOneTuple(dc::DC const& dc) {
+    std::vector<Column::IndexType> all_cols = dc.GetColumnIndices();
+    for (size_t i = 0; i < data_.front().GetNumRows(); ++i) {
+        if (ContainsNullOrEmpty(all_cols, i)) continue;
+        std::vector<std::byte const*> tuple = GetTuple(i);
+        if (Eval(tuple, dc.GetPredicates())) return false;
+    }
+
+    return true;
+}
+
+bool DCVerifier::VerifyTwoTuples(dc::DC const& dc) {
+    std::vector<Column::IndexType> all_cols = dc.GetColumnIndices();
+    std::vector<Column::IndexType> eq_cols = dc.GetColumnIndicesWithOperator(
+            [](dc::Operator op) { return op.GetType() == dc::OperatorType::kEqual; });
+    std::vector<Column::IndexType> ineq_cols = dc.GetColumnIndicesWithOperator(
+            [](dc::Operator op) { return op.GetType() != dc::OperatorType::kEqual; });
+
+    std::vector<Point> points;
+    std::unordered_map<size_t, util::KDTree<Point>> hash;
+    for (size_t i = 0; i < data_.front().GetNumRows(); ++i) {
+        if (ContainsNullOrEmpty(all_cols, i)) continue;
+
+        std::vector<std::byte const*> tuple = GetTuple(i);
+        size_t key = HashTuple(tuple, eq_cols);
+        Tree& tree = hash[key];
+
+        auto [box, inv_box] = SearchRanges(dc, tuple);
+        std::vector<Point> search_res = tree.QuerySearch(box);
+        std::vector<Point> inv_search_res = tree.QuerySearch(inv_box);
+        if (!search_res.empty() or !inv_search_res.empty()) return false;
+
+        tree.Insert(MakePoint(tuple, ineq_cols, i + index_offset_));
+    }
+
+    return true;
+}
+
+bool DCVerifier::VerifyAllEquality(dc::DC const& dc) {
+    std::unordered_set<size_t> res_tuples;
+    std::vector<Column::IndexType> const eq_cols = dc.GetColumnIndices();
+    for (size_t i = 0; i < data_.front().GetNumRows(); ++i) {
+        if (ContainsNullOrEmpty(eq_cols, i)) continue;
+        std::vector<std::byte const*> tuple = GetTuple(i);
+        size_t key = HashTuple(tuple, eq_cols);
+        if (res_tuples.find(key) != res_tuples.end()) return false;
+        res_tuples.insert(key);
```

Comment:
Yes, that would work fine

Reply:
If two different sets if values hash to the same number, this would falsely consider them equal and DC would be determined to not hold(?), even if it actually does (?).

Reply:
Hm, that actually may lead to hash collisions. We perhaps need to create a hashable structure that consists of `tuple` and `eq_cols` to be able to compare actual keys in hash table not simply their hash values

Reply:
Alternatively, write a custom hasher using the function you already have and plug it into the standard `std::unordered_set` as a template parameter